### PR TITLE
 Complete Activity by ID

### PR DESCRIPTION
**Pull request by vancexu**

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -102,6 +102,21 @@ type (
...
+		// should be called when that activity is completed with the actual result and error. If err is nil, activity task
+		// completed event will be reported; if err is CanceledError, activity task cancelled event will be reported; otherwise,
+		// activity task failed event will be reported.
+		// An activity implementation should use activityID provided in ActivityOption to use for completion.
+		// domainId, workflowId, activityId are required
+		// The errors it can return:
+		//  - ErrorWithDetails
+		//  - TimeoutError
+		//  - CanceledError
+		CompleteActivityByID(ctx context.Context, domainID, workflowID, activityID string, result interface{}, err error) error
```
Comment (mfateev) [inline client.go]: Add runID as an optional argument.

File: internal_task_pollers.go [hunk 2]
```diff
@@ -29,6 +29,7 @@ import (
 	"time"
 
 	"github.com/uber-go/tally"
+	common2 "github.com/uber/cadence/common"
```
Comment (mfateev) [inline internal_task_pollers.go]: rename to more meaningful: cadencecommon or ccommon

File: internal_task_pollers.go [hunk 3]
```diff
@@ -587,3 +588,22 @@ func convertActivityResultToRespondRequest(identity string, taskToken, result []
 		Details:   details,
 		Identity:  common.StringPtr(identity)}
 }
+
+func convertActivityResultToRespondRequestByID(identity, domainID, workflowID, activityID string,
```
Comment (mfateev) [inline internal_task_pollers.go]: this should call a correspondent service API. The token should be opaque to the clients.

File: client.go (hunk 1 above)
Comment (vancexu) [reply client.go]: Done

File: internal_task_pollers.go (hunk 2 above)
Comment (vancexu) [reply internal_task_pollers.go]: Removed since not used

File: internal_task_pollers.go (hunk 3 above)
Comment (vancexu) [reply internal_task_pollers.go]: Updated

**Comment by vancexu** [comment]
Server side changes see https://github.com/uber/cadence/pull/434

File: idl/github.com/uber/cadence/cadence.thrift [hunk 4]
```diff
@@ -174,6 +174,20 @@ service WorkflowService {
       3: shared.EntityNotExistsError entityNotExistError,
     )
 
+  /**
+  * RespondActivityTaskCompletedByID is called by application worker when it is done processing an ActivityTask.
+  * It will result in a new 'ActivityTaskCompleted' event being written to the workflow history and a new DecisionTask
+  * created for the workflow so new decisions could be made.  Use the 'taskToken' provided as response of
```
Comment (mfateev) [inline idl/github.com/uber/cadence/cadence.thrift]: This comment references taskToken. It looks like it was just copied from RespondActivityTaskCompleted

File: idl/github.com/uber/cadence/cadence.thrift [hunk 5]
```diff
@@ -188,6 +202,20 @@ service WorkflowService {
       3: shared.EntityNotExistsError entityNotExistError,
     )
 
+  /**
+  * RespondActivityTaskFailedByID is called by application worker when it is done processing an ActivityTask.
+  * It will result in a new 'ActivityTaskFailed' event being written to the workflow history and a new DecisionTask
+  * created for the workflow instance so new decisions could be made.  Use the 'taskToken' provided as response of
```
Comment (mfateev) [inline idl/github.com/uber/cadence/cadence.thrift]: Ditto

File: idl/github.com/uber/cadence/cadence.thrift [hunk 6]
```diff
@@ -202,6 +230,20 @@ service WorkflowService {
       3: shared.EntityNotExistsError entityNotExistError,
     )
 
+  /**
+  * RespondActivityTaskCanceledByID is called by application worker when it is successfully canceled an ActivityTask.
+  * It will result in a new 'ActivityTaskCanceled' event being written to the workflow history and a new DecisionTask
```
Comment (mfateev) [inline idl/github.com/uber/cadence/cadence.thrift]: Ditto

File: client.go [hunk 7]
```diff
@@ -102,6 +102,21 @@ type (
 		// The activity can fail with below errors ErrorWithDetails, TimeoutError, CanceledError.
 		CompleteActivity(ctx context.Context, taskToken []byte, result interface{}, err error) error
 
+		// CompleteActivityById reports activity completed.
+		// Similar to CompleteActivity, but may save cadence user from keeping taskToken info
```
Comment (mfateev) [inline client.go]: Fix the comment as it looks like multiple sentences without dots. The generated documentation removes all new lines (unless double new line).

File: idl/github.com/uber/cadence/cadence.thrift (hunk 4 above)
Comment (vancexu) [reply idl/github.com/uber/cadence/cadence.thrift]: Thanks for catching that! Updated.

**Pull request merged**